### PR TITLE
AdaptivePoolingAllocator's Chunks Weak Ref-Counting

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

AdaptivePoolingAllocator's buffer allocation, while bumping the pointer in the current chunk, is uncontended, but given that ChunkByteBuf extends AbstractReferenceCountedByteBuf for convenience (leak detection, double-free detection, etc etc), it still uses atomic operations. Despite Apple M1/M2/... are pretty fast while performing atomic uncontended operations, this is not so true for x86 ones.

Modification:

Make ChunkByteBuf to implement its own "weaker" reference-count logic, compared to AbstractReferenceCountedByteBuf, which benefit of a double counter, not atomic on the borrowing part.

Result:

Better allocation performance

### CODE REVIEW

**Comment by franz1981** [comment]
The interesting takes here:
- I can make this "weak" logic a bit less "weak" with some love (eg marking some final state on release, to avoid double-free)
- the `Chunk` is already wrapping a `ByteBuf` which has its own ref-counting (and leak tracking, in case)
- the `Chunk`s acquire operation(s) are all "internal" (ie not exposed to user code) and the release are wrapping `ByteBuf`'s releases or are part of the pool logic 

In short, i'm not quite sure of the value of our "classic" ref cnt scheme here, but despite you have an Apple machine (which is hell fast on uncontended atomics), I would be happy to see how it performs...let me know! @normanmaurer @chrisvest 

**NOTE**; the acquired buffer release still suffer from double ref-cnt (on the used `ByteBuf` and the new weaker scheme at `Chunk`).

**Review by chrisvest — APPROVED** [review]

**Comment by normanmaurer** [comment]
@franz1981 thanks a lot! Can you port this to main ?

**Comment by chrisvest** [comment]
`main` works differently in this area, so I'm not sure this applies.
